src/mono/dlls/mscordbi/cordb.cpp

```diff
@@ -483,6 +483,11 @@ void Connection::LoopSendReceive()
          MINOR_VERSION, major_version, minor_version));
     free(vm_version);
 
+    m_dbgprot_buffer_init(&localbuf, 128);
+    m_dbgprot_buffer_add_int(&localbuf, POS_RSP);
+    cmdId = SendEvent(MDBGPROT_CMD_SET_VM, MDBGPROT_CMD_VM_SET_CLIENT_POS_CONTEXT_SP, &localbuf);
+    m_dbgprot_buffer_free(&localbuf);
```

Comment:
This probably needs an explanation in a comment somewhere...

And actually... what is it doing? 

Here is what I understand: 

The VS debugger has an idea of what a native stack frame looks like.  And it expects to be able to decode it and inspect it.  But most importantly it expects to find the stack pointer value at an offset that it knows within the frame.

So when the debugger starts, we send that offset over to mono.  mono then uses it when it receives a "set context" command from the debugger to find the right stack frame?

(what I don't understand is why the debugger protocol "set context" command doesn't do anything - it just finds the index of the frame that has that stack pointer and returns it to the debugger.  I guess the debugger is setting something when it receives that?)

is that right?

Does the debugger know about other platforms?  Does it know their calling conventions - are we going to have to teach it about iOS/arm64?  Is it actually trying to do its own stack walking?  That seems very intrusive.

Reply:
-> (what I don't understand is why the debugger protocol "set context" command doesn't do anything - it just finds the index of the frame that has that stack pointer and returns it to the debugger. I guess the debugger is setting something when it receives that?)
SET_CONTEXT as I undertood is just used to set the current frame in the StackWalk, nothing is set on mono side, but is set on mscordbi side, all the information asked after this set_context are about this selected frame.

->Does the debugger know about other platforms? Does it know their calling conventions - are we going to have to teach it about iOS/arm64? Is it actually trying to do its own stack walking? That seems very intrusive.
As I could see the context is only used to check the caller SP and the callee SP to identify if there are more frames to walk. We woudn't need to teach iOS, my idea was use the client format of the context, and they know all the client formats: x86, amd64, arm64 and arm.

-> If the debugger only cares about the stack pointer, do we have to send the whole buffer for SET_CONTEXT? can we filter on the debugger side and only send the SP value to Mono?
Yes, I can change this.

Reply:
> We woudn't need to teach iOS, my idea was use the client format of the context, and they know all the client formats: x86, amd64, arm64 and arm.

@thaystg I didn't understand this part.  What I am concerned about is:  on apple devices the arm64 ABI is not the same as arm64 windows or linux.  So I just want to check that all the knowledge about the ARM64 ABI is in mscordbi or in mono - not in vsdbg.

Reply:
For vsdbg the SP will be in the position of the architecture where the debugger client is running. 
So for example debugger client is running on ARM64 windows, we will receive the expected position of stack pointer on windows using this message MDBGPROT_CMD_VM_SET_CLIENT_POS_CONTEXT_SP, on mono runtime we will get the information from MonoContext, it doesn't matter if it's running on iOS or Android or Windows because MonoContext has the knowledge, and we will fill the information on the position that the debugger client expects.

Reply:
Ok, that makes sense.  Thanks!